File: string-algorithms/assignment3/kmp/c/kmp.c
#include <stdio.h>
#include <stdlib.h>
#include "prefix.h"
int find_pattern(int pattern_len, char pattern[], int text_len, char text[],
                 int **matches) {
  int total_len = pattern_len + 1 + text_len;
  char pattern_and_text[total_len];

  int prefix_array[2000002];

  for (int i = 0; i < pattern_len; ++i) {
    pattern_and_text[i] = pattern[i];
  }

  pattern_and_text[pattern_len] = '$';

  for (int i = 0; i < text_len; ++i) {
    pattern_and_text[pattern_len + i + 1] = text[i];
  }

  prefix(total_len, pattern_and_text, prefix_array);

  int result_count = 0;

  for (int i = 0; i < total_len; ++i) {
    if (prefix_array[i] == pattern_len) {
      result_count++;
    }
  }

  *matches = malloc(result_count * (sizeof(int)));

  int match_pointer = 0;
  for (int i = 0; i < total_len; ++i) {
    if (prefix_array[i] == pattern_len) {
      (*matches)[match_pointer] = (i - (2 * pattern_len));
      match_pointer++;
    }
  }

  return result_count;
}

File: string-algorithms/assignment3/kmp/ruby/kmp.rb
def prefix(string)
  prefix_array = Array.new(string.length)
  prefix_array[0] = 0
  border = 0

  (1..string.length).each do |i|
    while ((border > 0) && (string[i] != string[border]))
      border = prefix_array[border - 1];
    end

    if (string[i] == string[border])
      border = border + 1;
    else
      border = 0;
    end

    prefix_array[i] = border;
  end

  return prefix_array
end

def find_pattern(pattern, text)
  total = pattern + "$" + text

  prefix_array = prefix(total)

  matches = []

  prefix_array.each_with_index do |item, index|
    if item == pattern.length
      matches.push(index - (2 * pattern.length))
    end
  end

  return matches
end

def main
  pattern = gets.chomp
  text = gets.chomp

  results = find_pattern(pattern, text)

  puts results.join(" ")
end

main